content/blog/2022/2022-05-27-one-stop-guide-mapstruct.md

```diff
@@ -0,0 +1,2116 @@
+        }
+
+        return list;
+    }
+    
+    ...
+}
+```
+
+#### Map
+
+MapStruct provides additional annotation for mapping Maps. It is annotated as `MapMapping` and it accepts custom definitions to define various formats for key-value pairs:
+
+```java
+@Mapper
+public interface CollectionMapper {
+    CollectionMapper INSTANCE = Mappers.getMapper(CollectionMapper.class);
+
+    @MapMapping(keyNumberFormat = "#L", valueDateFormat = "dd.MM.yyyy")
+    Map<String, String> map(Map<Long, Date> dateMap);
+}
+```
+
+This would generate an automated implementation method something like below:
+
+```java
+@Generated(
+    value = "org.mapstruct.ap.MappingProcessor"
+)
+public class CollectionMapperImpl implements CollectionMapper {
+    
+    ...
+    @Override
+    public Map<String, String> map(Map<Long, Date> dateMap) {
+        if ( dateMap == null ) {
+            return null;
+        }
+
+        Map<String, String> map = new HashMap<String, String>(
+            Math.max( (int) ( dateMap.size() / .75f ) + 1, 16 ) );
+
+        for ( java.util.Map.Entry<Long, Date> entry : dateMap.entrySet() ) {
+            String key = new DecimalFormat( "#L" ).format( entry.getKey() );
+            String value = new SimpleDateFormat( "dd.MM.yyyy" )
+                .format( entry.getValue() );
+            map.put( key, value );
+        }
+
+        return map;
+    }
+    
+    ...
+}
+
+```
+
+#### Mapping Strategies
+
+In case, if we need to map data types with the parent-child relationship, then MapStruct offers a way to define a strategy to set or add the children to the parent type. The `@Mapper` annotation supports a `collectionMappingStrategy` attribute which takes the following enums:
+
+* `ACCESSOR_ONLY`
+* `SETTER_PREFERRED`
+* `ADDER_PREFERRED`
+* `TARGET_IMMUTABLE`
+
+The default value is `ACCESSOR_ONLY`, which means that only accessors can be used to set the *Collection* of children. This option helps us when the adders for a Collection type field are defined instead of setters. For example, let’s revisit the `Manager` to `ManagerDO` entity conversion in `PersonDTO`. The `PersonDTO` entity has a child field of type `List`:
+
+```java
+public class PersonDTO {
+    ...
+    private List<ManagerDTO> managerList;
+    
+    public List<ManagerDTO> getManagerList() {
+        return managers;
+    }
+
+    public void setManagerList(List<ManagerDTO> managers) {
+        this.managers = managers;
+    }
+
+    public void addManagerList(ManagerDTO managerDTO) {
+        if (managers == null) {
+            managers = new ArrayList<>();
+        }
+
+        managers.add(managerDTO);
+    }
+    
+    // other getters and setters
+}
+```
+
+Note that we have both the setter method, `setManagers`, and the adder method, `addManagerList` and we are responsible to initiate the collection for the adder. Then we have defined the default mapper the implementation looks something like the below:
+
+```java
+@Generated(
```

Comment:
This is a case where showing the generated code makes a lot of sense!

Reply:
Agreed!